Client Report Implementation

DedrickEnc opened the pull request:
This PR implements the client report and it is in progress as checkbox are showing

- [x]  Client side implementation 

- [x] Server side :  data processing

- [x] Server side : template implementation

- [x] Cleaning the report 

- [x]  testing

closes #813

DedrickEnc committed:
client side, data fetching and structuring

DedrickEnc committed:
templating done

DedrickEnc committed:
report done

DedrickEnc committed:
integration test

DedrickEnc committed:
update test

sfount commented on client/src/partials/reports/modals/clients_report.config.js:
```diff
@@ -0,0 +1,59 @@
+      })
+      .catch(function (error) {
+        vm.$loading = false;
+        throw error;
```
This will show nothing to the user if something goes wrong. If something should be displayed `Notify` should be used as in reading Clients data.

sfount commented on server/controllers/finance/reports/clientsReport/index.js:
```diff
@@ -0,0 +1,201 @@
+
+  if(session.ignoredClients){
+    ignoredClients = (Array.isArray(session.ignoredClients))? session.ignoredClients : [session.ignoredClients];
+    ignoredToken = ` AND dg.uuid NOT IN (${escapeItems(ignoredClients).join(',')})`;
```
Using an array and ? syntax 
```uuid NOT IN ?``` 
usually produces a comma separated list - is there a specific need to treat this array differently?

DedrickEnc replied:
Angular sends an array if there are many client to ignore and a simple string if there is only one client to ignore, that why I used that.

sfount commented on server/controllers/finance/reports/clientsReport/report.handlebars:
```diff
@@ -0,0 +1,55 @@
+    </thead>
+
+    <tbody>
+      {{#annualLine lines detailPrevious enterprise.currency_id}}
```
Can the built in iterator be used instead of these complex helpers? 

Something like:
```html
{{#each lines}}
<tr>
<td>{{this.accountNumber}}<td>
<td>{{this.name}}</td>
{{#if detailPrevious}}
<td>{{this.initDebit}}</td>
<td>{{this.initCredit}}</td>
{{/if}}
</tr>
```

DedrickEnc replied:
I am iterating on an object properties, not on an array that why I have created the helper

sfount replied:
The built in `#each` helper iterates over objects, as in the example shown. 
http://handlebarsjs.com/builtin_helpers.html

sfount reviewed (commented):
The code here looks okay to me. 

I have made a few comments on complex solutions that might be simplified by using standard practices ('?' syntax and build in handlebars helpers). 

I also have a question on the functionality - is it necessary to have a `from` and a `to` date on something representing a balance? or is it just useful to see the balance up to that date? 
This report is very similar to the aged debtor report - we should make sure that we are very clear what each report is used for to not confuse users. They may even be able to share code for maintainability reasons.

DedrickEnc committed:
applying suggestion

DedrickEnc committed:
fixing conflicts

DedrickEnc committed:
some fixes on client report

DedrickEnc committed:
Refactor the clients report

Move all mathematical operation to the server side
Use bhima translate directive to the view

Notice that the total balance for each cleint is calculated in nodejs,
with is a addition of the balance of previous year and the balance of the
currente year.
The lodash library was used to make simple some assignation or affectation

Resolves : #813

sfount commented on server/controllers/finance/reports/clientsReport/report.handlebars:
```diff
@@ -0,0 +1,68 @@
+          <td>{{this.name}}</td>
+          {{#if ../detailPrevious}}
+            <td class='text-right'> {{currency this.initDebit enterprise.currency_id}} </td>
+            <td class='text-right'> {{currency this.initCredit enterprise.currency_id}} </td>
```
I don't think this has access to enterprise as it is in the each loop - in my experience you have to use `../enterprise.currency_id` to get to the top object. 

This may be a special case but otherwise this evening will always use a default currency.

DedrickEnc replied:
The enterprise currency is accessed and the report is rendered correctly may be I don't understand the question

DedrickEnc replied:
You are right @sfount

sfount reviewed (commented):
This looks like it has well factored in my comments, I have one question making sure the handlebars template works but then this can be merged!
